lib/IMAP/MessageMapper.php

```diff
@@ -75,40 +77,94 @@ public function find(Horde_Imap_Client_Base $client,
 	 *
 	 * @param int|null $highestKnownUid
 	 *
-	 * @return IMAPMessage[]
+	 * @return array
 	 * @throws Horde_Imap_Client_Exception
 	 */
 	public function findAll(Horde_Imap_Client_Socket $client,
 							Mailbox $mailbox,
-							int $highestKnownUid = null,
-							int $maxResults = null): array {
+							int $maxResults,
+							?int $highestKnownUid = 0): array {
+		/**
+		 * To prevent memory exhaustion, we don't want to just ask for a list of
+		 * all UIDs and limit them client-side. Instead we can (hopefully
+		 * efficiently) query the min and max UID as well as the number of
+		 * messages. Based on that we assume that UIDs are somewhat distributed
+		 * equally and build a page to fetch.
+		 *
+		 * This logic might return fewer or more results than $maxResults
+		 */
+
+		$metaResults = $client->search(
+			$mailbox->getName(),
+			null,
+			[
+				'results' => [
+					Horde_Imap_Client::SEARCH_RESULTS_MIN,
+					Horde_Imap_Client::SEARCH_RESULTS_MAX,
+					Horde_Imap_Client::SEARCH_RESULTS_COUNT,
+				]
+			]
+		);
+		/** @var int $min */
+		$min = $metaResults['min'];
+		/** @var int $max */
+		$max = $metaResults['max'];
+		/** @var int $total */
+		$total = $metaResults['count'];
+
+		if ($total === 0) {
+			// Nothing to fetch for this mailbox
+			return [
+				'messages' => [],
+				'all' => true,
+			];
+		}
+
+		// The inclusive range of UIDs
+		$totalRange = $max - $min + 1;
+		// Here we assume somewhat equally distributed UIDs
+		// +1 is added to fetch all messages with the rare case of strictly
+		// continuous UIDs and fractions
+		$estimatedPageSize = (int)(($totalRange / $total) * $maxResults) + 1;
```

Comment:
```suggestion
		$estimatedPageSize = (int) (($totalRange / $total) * $maxResults) + 1;
```

You use space in the file above, I'm fine without, but pick one style :grin: